## [Νικόλαος Μανιάτης](mailto:nikolaosmaniatis@mail.ntua.gr)
## Α.Μ.: 03400097

<h1>Q-Learning</h1>

<p>Στην συγκεκριμένη άσκηση θα μελετήσετε το στοχαστικό αλγόριθμο Q-Learning, χρησιμοποιώντας το έτοιμο πρόγραμμα που σας δίνεται.</p> <p> Στα πλαίσια του παραδείίγματος θα εξετάσετε μίία υλοποίηση με q-learning σχετικά με το σύστημα αυτόματης οδήγησης ενός ταξί. Στα πλαίσια του προβλήματος αυτού θα πρέπει να ικανοποιούνται τα εξής:</p>
<ul>
<li>Το ταξί θα πρέπει να αφήνει τον πελάτη στη σωστή θέση</li>
<li>Το ταξί να ακολουθεί τη συντομότερη δυνατή διαδρομή</li>
<li>Να τηρούνται οι κανόόνες κυκλοφορίας και ασφάλειας των επιβατών</li>
</ul>

<p>Στα πλαίσια του προβλήματος έχουμε τα εξής χαρακτηριστικά για την επιβράβευση, τις καταστάσεις και τις ενέργειες.</p> 

<h4>Επιβράβευση</h4>
<ul>
<li>Θα έχουμε τη μέγιστη επιβράβευση όταν το ταξί αφήνει έναν πελάτη στην επιθυμητή θέση</li>
<li>Θα υπάρχει ποινή στην περίπτωση όπου το ταξί αφήσει τον πελάτη σε κάποιο λανθασμένο σημείο</li>
<li>Ο agent θα παίρνει μία μικρή σχετικά ποινή στην περίπτωση όπου αργεί να φτάσει στον τελικό προορισμό</li>
</ul>

<p>Γενικά οι παραπάνω αρχές συνοψίζονται στα εξής: "Λαμβάνουμε +20 πόντους για μια επιτυχημένη πτώση και χάνουμε 1 πόντο για κάθε χρονικό βήμα που παίρνει. Υπάρχει επίσης ποινή 10 πόντων για παράνομες ενέργειες παραλαβής και αποχώρησης."</p>

<h4>Πλήθος Καταστάσεων</h4>
<img src="https://storage.googleapis.com/lds-media/images/Reinforcement_Learning_Taxi_Env.width-1200.png">
<p>Στο παράδειγμά μας έχουμε ένα μικρό χώρο 5*5. Από εκεί και πέρα έχουμε 4 προορισμούς και 5 πιθανές θέσεις του πελάτη (παίρνουμε και την περίπτωση ο πελάτης να είναι ήδη μέσα στο τάξι).</p>
<p>Με βάση τα παραπάνω έχουμε 5*5*5*4=500 πιθανές καταστάσεις.</p>

<h4>Πλήθος Ενεργειών</h4>
<p>Έχουμε έξι ενεργειες για το ταξί, οι οποίες είναι οι εξής:</p>
<ul>
<li>0=Νότια</li>
<li>1=Βόρεια</li>
<li>2=Ανατολικά</li>
<li>3=Δυτικά</li>
<li>4=Επιβίβαση</li>
<li>5=Αποβίβαση</li>
</ul>





<p>Πριν συνεχίσετε στην άσκηση να απαντήσετε στο εξής ερώτημα:</p>

<b><p>1. Να περιγράψετε σύντομα τον αλγόριθμο Q-Learning. Σε ποια προβλήματα θεωρείτε ότι ταιριάζει ως τρόπος εκμάθησης η Ενισχυτική Μάθηση (Reinforcement Learning); Ποια είναι η βασική διαφορά του αλγορίθμου Q-Learning από τους αλγορίθμους Policy Iteration και Value Iteration;</p></b>

Ο αλγόριθμος Q-Learning λειτουργεί με το να ορίσουμε έναν πίνακα με την αξία κάθε ζεύγους state και action για το πρόβλημα που επιχειρούμε να λύσουμε, οποίος αρχικά είναι αρχικοποιημένος με μηδενικά, και γεμίζει με κάθε επανάληψη του αλγορίθμου. Υιοθετούμε μια $\varepsilon$-greedy πολιτική, όπου με πιθανότητα $1-\varepsilon$ ο αλγόριθμος εκτελεί την κίνηση που μεγιστοποιεί το κέρδος του (exploitation) ενώ με πιθανότητα $\varepsilon$ εκτελεί μια τυχαία κίνηση (exploration). Έπειτα, με βάση τη γνωστή σχέση του Q-Learning ανανεώνει τις τιμές του πίνακα και συνεχίζει τις επαναλήψεις. 
Η σχέση αυτή είναι:   

$$\mathcal{Q}^{new} (s_t,a_t) \leftarrow \mathcal{Q}(s_t,a_t) + \alpha(r_t + \gamma\cdot\max_a \mathcal{Q}(s_{t+1},a) - \mathcal{Q}(s_t,a_t)) $$ 

H βασική διαφορά του Q-Learning από τους αλγορίθμους Policy & Value Iteration είναι ότι ο αλγόριθμος Q-Learning δεν θεωρεί γνωστές τις πιθανότητες μεταβάσεων μεταξύ των καταστάσεων ούτε των rewards. Επομένως, με την βοήθεια προσωμοιώσεων Monte Carlo προσπαθούμε να βρούμε κάποια βέλτιστη πολιτική του προβλήματος που εξετάζουμε. 

<p>Στη συνέχεια θα πρέπει να φορτώσετε τη βιβλιοθήκη gym καθώς και το σχετικό dataset<p>

In [5]:
import gym
import numpy as np
env = gym.make("Taxi-v3").env

env.render()

+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



In [6]:
env.reset() # reset environment to a new, random state
env.render()

print("Action Space {}".format(env.action_space))
print("State Space {}".format(env.observation_space))


+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+

Action Space Discrete(6)
State Space Discrete(500)


<p>Παρακάτω ορίζουμε τις συνεταγμένες του ταξί, τη θέση του πελάτη και το σημείο προορισμού</p>

In [7]:
state = env.encode(3, 1, 2, 0) # (taxi row, taxi column, passenger index, destination index)
print("State:", state)

env.s = state
env.render()

State: 328
+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



<p>Παρακάτω είναι η μήτρα επιβράβευσης για το state που ορίσαμε στο προηγούμενο βήμα</p>

In [8]:
env.P[328]

{0: [(1.0, 428, -1, False)],
 1: [(1.0, 228, -1, False)],
 2: [(1.0, 348, -1, False)],
 3: [(1.0, 328, -1, False)],
 4: [(1.0, 328, -10, False)],
 5: [(1.0, 328, -10, False)]}

<p> Τρέχουμε το παράδειγμά μας χωρις τη χρήση Q-Learning.</p>

<b><p>2. Τα αποτελέσματα είναι ικανοποιητικά; Πως θα μας εξυπηρετούσε η χρήση του Q-Learning;</p></b>

In [14]:
env.s = 328  # set environment to illustration's state

epochs = 0
penalties, reward = 0, 0

frames = [] # for animation

done = False

while not done:
    action = env.action_space.sample()
    state, reward, done, info = env.step(action)

    if reward == -10:
        penalties += 1
    
    # Put each rendered frame into dict for animation
    frames.append({
        'frame': env.render(mode='ansi'),
        'state': state,
        'action': action,
        'reward': reward
        }
    )

    epochs += 1
    
    
print("Timesteps taken: {}".format(epochs))
print("Penalties incurred: {}".format(penalties))

Timesteps taken: 1051
Penalties incurred: 334


Όπως είναι προφανές τα αποτελέσματα αυτά δεν είναι ικανοποιητικά καθώς η επιλογή της επόμενης κατάστασης γίνεται ανεξάρτητα από τα rewards και τα penalties, γι'αυτό και περίπου το 1/3 των κινήσεων οδηγήθηκαν σε penalty.  
Η χρήση του Q-Learning θα μας ήταν πολύ χρήσιμη καθώς ο αλγόριθμος μαθαίνει να κινείται διαλέγοντας κινήσεις που μεγιστοποιούν το κέρδος του και επομένως αποφεύγουν τα penalties. 

<p>Τώρα θα προσπαθήσουμε να λύσουμε το πρόβλημά μας με τη χρήση του Q-Learning.</p>

<b><p>3. Τι γνωρίζετε για τις παραμέτρους α και γ. Τι θα συμβεί αν έχουν τιμές ίσες με 1;</p></b>

 - Η παράμετρος $\alpha$ είναι το Learning Rate, αν δούμε την σχέση του Q-Learning, εκφράζει κατά τι ποσοστό η νέα πληροφορία θα κάνει overwrite την παλιά πληροφορία. Αν $\alpha = 0$, τότε ο agent δεν μαθαίνει καθόλου, και οι τιμές του Q δεν ανανεώνονται καθόλου. Αν πάλι $\alpha = 1$ αυτό σημαίνει ότι μαθαίνει πολύ γρήγορα, και σε ένα στοχαστικό περιβάλλον π.χ. αυτό μπορεί να μην είναι επιθυμητό. Γι'αυτό συνήθως διαλέγουμε μια σταθερή τιμή όπως στο παράδειγμά μας όπου $\alpha = 0.1$ . 
 - Η παράμετρος $\gamma$ είναι το discount factor. Αυτό ορίζει την σημασία των μελλοντικών rewards που θα πάρει ο agent. Αν $\gamma = 0$, τότε ο agent είναι μύωπας, δηλαδή δεν φροντίζει καθόλου για το μέλλον και παίρνει όποια απόφαση μεγιστοποιεί το άμεσο κέρδος του. Αν πάλι $\gamma \rightarrow 1 $, η state-value function που μας δείχνει τα αναμενόμενα κέρδη όταν ακολουθούμε μια πολιτική, και δίνεται από την σχέση $ u_{\pi} = \mathrm{E}\big [ \displaystyle \sum_{k=0}^{\infty} \gamma^k R_{t+k+1} \big| s_t =s\big ]$, αποκλίνει, αν δεν υπάρχει τερματική κατάσταση. Επομένως, πρέπει να είναι $\gamma < 1$ προκειμένου ο αλγόριθμος να συγκλίνει.

In [101]:
import numpy as np
q_table = np.zeros([env.observation_space.n, env.action_space.n])

In [102]:
%%time
"""Training the agent"""

import random
from IPython.display import clear_output

# Hyperparameters
alpha = 0.1
gamma = 0.6
epsilon = 0.1

# For plotting metrics
all_epochs = []
all_penalties = []

for i in range(1, 100001):
    state = env.reset()

    epochs, penalties, reward, = 0, 0, 0
    done = False
    
    while not done:
        if random.uniform(0, 1) < epsilon:
            action = env.action_space.sample() # Explore action space
        else:
            action = np.argmax(q_table[state]) # Exploit learned values

        next_state, reward, done, info = env.step(action) 
        
        old_value = q_table[state, action]
        next_max = np.max(q_table[next_state])
        
        new_value = (1 - alpha) * old_value + alpha * (reward + gamma * next_max)
        q_table[state, action] = new_value

        if reward == -10:
            penalties += 1

        state = next_state
        epochs += 1
        
    if i % 100 == 0:
        clear_output(wait=True)
        print(f"Episode: {i}")

print("Training finished.\n")

Episode: 100000
Training finished.

CPU times: user 35.3 s, sys: 7.44 s, total: 42.8 s
Wall time: 33.6 s


In [103]:
q_table[328]

array([ -2.39346117,  -2.27325184,  -2.40801732,  -2.35679237,
       -10.521667  , -10.68555429])

In [104]:
"""Evaluate agent's performance after Q-learning"""

total_epochs, total_penalties = 0, 0
episodes = 100
rewardss = []
rewards_l = []
for _ in range(episodes):
    state = env.reset()
    epochs, penalties, reward = 0, 0, 0
    
    done = False
    
    while not done:
        action = np.argmax(q_table[state])
        state, reward, done, info = env.step(action)
        rewardss.append(reward)
        if reward == -10:
            penalties += 1

        epochs += 1

    total_penalties += penalties
    total_epochs += epochs
    rewards_l.append(np.mean(rewardss))
print(f"Results after {episodes} episodes:")
print(f"Average timesteps per episode: {total_epochs / episodes}")
print(f"Average penalties per episode: {total_penalties / episodes}")
print("Average rewards per step {:2f}".format(np.mean(rewards_l)))

Results after 100 episodes:
Average timesteps per episode: 13.49
Average penalties per episode: 0.0
Average rewards per step 0.579308


<b><p>4. Συγκρίνετε τους δύο αλγορίθμους με βάση τις παρακάτω μετρικές</p>
<ul>
<li>Μέσος αριθμός παραβάσεων ανά επεισόδιο</li>
<li>Μέσος αριθμός βημάτων ανά διαδρομή</li>
<li>Μέσος αριθμός ανταμοιβών ανά κίνηση</li>
</ul>
<p>Τις παραπάνω συγκρίσεις να τις κάνετε για 100 επεισόδια.</p>
</b>

In [72]:
def crude_version():
    env.s = 328  # set environment to illustration's state
    epochs = 0
    penalties, reward = 0, 0

    frames = [] # for animation

    done = False
    rewards = []
    while not done:
        action = env.action_space.sample()
        state, reward, done, info = env.step(action)
        rewards.append(reward)
        if reward == -10:
            penalties += 1

        # Put each rendered frame into dict for animation
        frames.append({
            'frame': env.render(mode='ansi'),
            'state': state,
            'action': action,
            'reward': reward
            }
        )

        epochs += 1
#     print("Timesteps taken: {}".format(epochs))
#     print("Penalties incurred: {}".format(penalties))
    return penalties,epochs,rewards

In [95]:
penalties_list = []
epochs_list = []
rewards_list = []
for i in range(100):
    penalties, epochs, rewards = crude_version()
    penalties_list.append(penalties)
    epochs_list.append(epochs)
    rewards_list.append(np.mean(rewards))

In [97]:
print('Average Number of penalties per Episode: {:.2f}'.format(np.mean(penalties_list)))
print('Average Number of steps per Episode: {:.2f}'.format(np.mean(epochs_list)))
print('Average Rewards per Step: {:.2f}'.format(np.mean(rewards_list)))

Average Number of penalties per Episode: 623.27
Average Number of steps per Episode: 1921.30
Average Rewards per Step: -3.85


### Αποτελέσματα (για 100 επεισόδια) : 
 - Χωρίς χρήση Q-Learning:   
   Μέσος αριθμός παραβάσεων ανά επεισόδιο: 623.27   
   Μέσος αριθμός βημάτων ανά διαδρομή (επεισόδιο): 1921.30   
   Μέσος αριθμός ανταμοιβών ανά κίνηση: -3.85   
   
   --------------
   
 - Με Q-Learning:   
   Μέσος αριθμός παραβάσεων ανά επεισόδιο: 0.0    
   Μέσος αριθμών βημάτων ανά διαδρομή (επεισόδιο): 13.49    
   Μέσος αριθμός ανταμοιβών ανά κίνηση: 0.58    

Είναι εμφανές ότι ο αλγόριθμος Q-Learning φέρνει εντυπωσιακά αποτελέσματα σε σχέση με τον απλό αλγόριθμο που είναι random, καθώς ο Q-Learning έχει την ικανότητα να μαθαίνει από την εκπαίδευσή του, και στο inference έχει μάθει για κάθε κατάσταση ποιες είναι οι ιδανικές ενέργειες, και έτσι καταφέρνει και μεγιστοποιεί το κέρδος του, εκμηδενίζοντας τον αριθμό των ποινών.